<a href="https://colab.research.google.com/github/ritika-33/image-classification/blob/main/catndogimgclassification.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!pip install kaggle

In [ ]:
from google.colab import files
files.upload()

In [4]:

! mkdir ~/.kaggle

mkdir: cannot create directory ‘/root/.kaggle’: File exists


In [5]:
! cp kaggle.json ~/.kaggle/

In [6]:
! chmod 600 ~/.kaggle/kaggle.json

In [7]:
! kaggle datasets download -d samuelcortinhas/cats-and-dogs-image-classification #importing dataset

Dataset URL: https://www.kaggle.com/datasets/samuelcortinhas/cats-and-dogs-image-classification
License(s): CC0-1.0
 88% 57.0M/64.4M [00:00<00:00, 83.6MB/s]
100% 64.4M/64.4M [00:00<00:00, 84.3MB/s]


In [8]:
!unzip cats-and-dogs-image-classification.zip

Archive:  cats-and-dogs-image-classification.zip
  inflating: test/cats/cat_1.jpg     
  inflating: test/cats/cat_106.jpg   
  inflating: test/cats/cat_109.jpg   
  inflating: test/cats/cat_113.jpg   
  inflating: test/cats/cat_114.jpg   
  inflating: test/cats/cat_116.jpg   
  inflating: test/cats/cat_118.jpg   
  inflating: test/cats/cat_119.jpg   
  inflating: test/cats/cat_124.jpg   
  inflating: test/cats/cat_140.jpg   
  inflating: test/cats/cat_147.jpg   
  inflating: test/cats/cat_156.jpg   
  inflating: test/cats/cat_158.jpg   
  inflating: test/cats/cat_162.jpg   
  inflating: test/cats/cat_18.jpg    
  inflating: test/cats/cat_190.jpg   
  inflating: test/cats/cat_203.jpg   
  inflating: test/cats/cat_223.jpg   
  inflating: test/cats/cat_234.jpg   
  inflating: test/cats/cat_244.jpg   
  inflating: test/cats/cat_251.jpg   
  inflating: test/cats/cat_255.jpg   
  inflating: test/cats/cat_268.jpg   
  inflating: test/cats/cat_279.jpg   
  inflating: test/cats/cat_281.jpg   
 

In [10]:
!pip install keras

In [11]:
!pip install tensorflow
from tensorflow.keras.preprocessing.image import ImageDataGenerator

In [38]:
import tensorflow
from tensorflow import keras
from keras import Sequential
from keras.layers import Dense, Flatten
from keras.applications.vgg16 import VGG16

In [14]:
conv_base =VGG16(weights='imagenet',
                 include_top=False,
                 input_shape=(150,150,3)
                 )

58889256/58889256 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step


## **Fine Tuning**

In [35]:
conv_base.trainable=True

In [37]:
set_trainable=False
for layer in conv_base.layers:
  if layer.name=='block5_conv1':
    set_trainable=True
  if set_trainable:
    layer.trainable=True
  else:
    layer.trainable=False
for layer in conv_base.layers:
  print(layer.name, layer.trainable)

input_layer False
block1_conv1 False
block1_conv2 False
block1_pool False
block2_conv1 False
block2_conv2 False
block2_pool False
block3_conv1 False
block3_conv2 False
block3_conv3 False
block3_pool False
block4_conv1 False
block4_conv2 False
block4_conv3 False
block4_pool False
block5_conv1 True
block5_conv2 True
block5_conv3 True
block5_pool True


In [ ]:
conv_base.summary()

In [ ]:
model =Sequential()
model.add(conv_base)
model.add(Flatten())
model.add(Dense(256,activation='relu'))
model.add(Dense(1,activation='sigmoid'))

In [19]:
batch_size= 16 #data aumentation
train_datagen = ImageDataGenerator(rescale=1./255,
                                   shear_range=0.2,
                                   zoom_range=0.2,
                                   horizontal_flip=True)
test_datagen = ImageDataGenerator(rescale=1./255)
train_generator=train_datagen.flow_from_directory('/content/train',
                                                  target_size=(150,150),
                                                  batch_size=batch_size,
                                                  class_mode='binary')
validation_generator=test_datagen.flow_from_directory('/content/test',
                                                      target_size=(150,150),
                                                      batch_size=batch_size,
                                                      class_mode='binary')

Found 557 images belonging to 2 classes.
Found 140 images belonging to 2 classes.


Below line of code prepare model for training by specifying how it will learn and optimize its performance during the training process using the RMSprp optimizer

In [20]:
model.compile(optimizer=keras.optimizers.RMSprop(learning_rate=1e-5),loss='binary_crossentropy',metrics=['accuracy'])

In [23]:
history=model.fit(train_generator,
                  steps_per_epoch=100,
                  epochs=10,
                  validation_data=validation_generator)

Epoch 1/10
 35/100 ━━━━━━━━━━━━━━━━━━━━ 4:58 5s/step - accuracy: 0.9529 - loss: 0.1609

/usr/lib/python3.10/contextlib.py:153: UserWarning: Your input ran out of data; interrupting training. Make sure that your dataset or generator can generate at least `steps_per_epoch * epochs` batches. You may need to use the `.repeat()` function when building your dataset.
  self.gen.throw(typ, value, traceback)


100/100 ━━━━━━━━━━━━━━━━━━━━ 197s 2s/step - accuracy: 0.9497 - loss: 0.1565 - val_accuracy: 0.7857 - val_loss: 0.3861
Epoch 2/10
100/100 ━━━━━━━━━━━━━━━━━━━━ 197s 2s/step - accuracy: 0.9432 - loss: 0.1484 - val_accuracy: 0.7643 - val_loss: 0.4122
Epoch 3/10
100/100 ━━━━━━━━━━━━━━━━━━━━ 197s 2s/step - accuracy: 0.9750 - loss: 0.1035 - val_accuracy: 0.8000 - val_loss: 0.3757
Epoch 4/10
100/100 ━━━━━━━━━━━━━━━━━━━━ 201s 2s/step - accuracy: 0.9705 - loss: 0.0946 - val_accuracy: 0.8286 - val_loss: 0.4006
Epoch 5/10
100/100 ━━━━━━━━━━━━━━━━━━━━ 198s 2s/step - accuracy: 0.9814 - loss: 0.0807 - val_accuracy: 0.8143 - val_loss: 0.4238
Epoch 6/10
100/100 ━━━━━━━━━━━━━━━━━━━━ 197s 2s/step - accuracy: 0.9906 - loss: 0.0704 - val_accuracy: 0.8000 - val_loss: 0.4190
Epoch 7/10
100/100 ━━━━━━━━━━━━━━━━━━━━ 198s 2s/step - accuracy: 0.9901 - loss: 0.0568 - val_accuracy: 0.8286 - val_loss: 0.4605
Epoch 8/10
100/100 ━━━━━━━━━━━━━━━━━━━━ 201s 2s/step - accuracy: 0.9837 - loss: 0.0607 - val_accuracy: 0.814

In [27]:
from sklearn.metrics import confusion_matrix, f1_score
import numpy as np

In [28]:
validation_steps = len(validation_generator)  #calculating predicted classes
predictions = model.predict(validation_generator, steps=validation_steps)
predicted_classes = (predictions > 0.5).astype(int)  # Convert probabilities to class labels (0 or 1)

9/9 ━━━━━━━━━━━━━━━━━━━━ 35s 4s/step


In [29]:
true_classes = validation_generator.classes  # Calculatig true classes to retrives the actual labels associated with validation dataset

In [31]:
cm = confusion_matrix(true_classes, predicted_classes)#Confusion Matrix
print("Confusion Matrix:")
print(cm)

Confusion Matrix:
[[37 33]
 [34 36]]


In [32]:
f1 = f1_score(true_classes, predicted_classes)# F1 SCORE
print("F1 Score:", f1)

F1 Score: 0.5179856115107914


Conclusion

This project aimed to develop an image classification model capable of distinguishing between images of cats and dogs. Using a subset of the Kaggle "Cats and Dogs Image Classification" dataset and the pre-trained VGG16 convolutional neural network architecture within the Google Colab environment, the project successfully trained a model for this task.The project utilized the VGG16 model pre-trained on the ImageNet dataset as a base, leveraging transfer learning to accelerate the training process. The model was fine-tuned by unfreezing certain layers and training them specifically on the cats and dogs dataset. Crucially, data augmentation techniques, including random image shearing, zooming, and horizontal flipping, were applied to artificially expand the training dataset. This approach aimed to increase the diversity of the training data, thereby improving the model's generalization ability and robustness to variations in real-world images.The performance of the trained model was assessed using metrics such as the F1 score. The achieved F1 score of approximately 0.52 indicates that the model has moderate classification ability. While exceeding random chance, it reveals areas for potential improvement. The confusion matrix provided further insights into the model's performance, highlighting specific types of classification errors.